### ISSUE TITLE
CompCert as a library, building CompCert with -Q

### ISSUE THREAD

**Issue by andrew-appel** [AbsInt/CompCert#199]
When CompCert is used as a library imported into other Coq projects, it is a problem that CompCert is built with a loadpath that uses -R instead of -Q.  

The front end of CompCert, up to and including common/Clight.v and exportclight/Clightdefs.v, is used as a library imported into VST and perhaps other projects.   Now that VST is also used as a library imported into arbitrary user projects, we want to make sure that VST does not pollute the namespace of these other projects.  That means we build VST with -Q in the loadpath instead of -R.

However, the user of VST must also import the front end of CompCert, up to Clightdefs.  If that is built with -R, then all the names defined in (the front end of) CompCert get imported into the user's project.

I propose two solutions to this problem.
1.  Build CompCert with -Q instead of -R.   That is, change the COQINCLUDES in the Makefile to use -Q instead of -R.  This requires changing many "Require" statements in the .v files, but that should be fairly straightforward.

2.  Alternate hack, which would improve the situation from the current one, but is imperfect:
Build CompCert with -R, as it is now, but import CompCert into other projects using -Q.
This hack requires minor changes to Clightdefs.v and ExportClight.ml.
The problem with this hack is that the Module namespace is still polluted.

In either case, Clightdefs.v must be modified as follows:
* All the Require statements should use fully qualified paths.
* Do not "Require Export" in the Clightdefs.v  (that pollutes the namespace for anyone that has to import Clightdefs).  I recommend changing all the Export to Import, but this also requires that ExportClight.ml print one additional line after it prints "Require Import Clightdefs", that is: 
"Import Coqlib Integers Floats AST Ctypes Cop Clight."

The reason that solution 2 still pollutes the namespace (slightly) is that "Clight.Tpointer" is now visible to anyone who imports Clightdefs.v, instead of "compcert.common.Clight.Tpointer".  But I believe (if I understand correctly) that in solution 1, a user could Import compcert.exportclight.Clightdefs, and then the shortest path to Tpointer would be compcert.common.Clight.Tpointer.

**Comment by andrew-appel** [comment]
There will be one issue to resolve with solution 1:  $(ARCH).  In the Makefile it's easy enough to make that a defined name, but a fully qualified Require statement will need a static name.

One solution that might work is:  omit $(ARCH) from RECDIRS, instead special-case it in the COQINCLUDES as,   -Q $(ARCH)  compcert.architecture.

There is also the issue of Flocq -- it should also be built with -Q.

**Comment by xavierleroy** [comment]
I need to be reminded what `-R` and `-Q` do and when to use one or the other.  Is there a good explanation somewhere?

**Comment by andrew-appel** [comment]
Coq reference manual, section 14.3.3:

     -Q directory dirpath
     Add physical path directory to the list of directories where Coq looks for a file and bind it to the logical directory dirpath. The subdirectory structure of directory is recursively available from Coq using absolute names (extending the dirpath prefix) (see Section 2.6.2).

     -R directory dirpath
     Do as -Q directory dirpath but make the subdirectory structure of directory recursively visible so that the recursive contents of physical directory is available from Coq using short or partially qualified names.

[end quote]
The "-Q" option was added only about two or three years ago; that's a good explanation of why CompCert 1.x did not use it!

In summary:  if you import with -Q, your Require statements (etc.) must use fully qualified names, e.g., compcert.cfrontend.Clight.   

When another big project imports compcert or VST as a library, it does not want its namespace polluted with all the unqualified names from compcert or VST.  Therefore, building with -Q is more polite.

**Comment by xavierleroy** [comment]
Thanks for the additional explanations.

I would be happier with putting all of CompCert's module in a `compcert.` namespace, e.g. `compcert.Clight` and `compcert.Asmgen` and `compcert.Archi`, regardless of subdirectory.  I'm afraid this isn't possible?

**Comment by andrew-appel** [comment]
You could try something like,
`-Q cfrontend compcert -Q common compcert -Q x86 compcert`
and see if it works.

**Comment by bschommer** [comment]
One problem: The menhir generated parser seems to have hardcoded `Require Import ...` directives for `Grammar`, `Automaton` etc.

**Comment by bschommer** [comment]
The rest seems to work relatively fine.

**Comment by jhjourdan** [comment]
About the menhir problem: This requires a patch to menhir, which is not a big deal.

I propose to make the menhir support libraries live in a dedicated namespace (say `menhir.Grammar`, etc...). This makes more sense, since they are possibly used by other projects.

**Comment by bschommer** [comment]
I patched my menhir locally with a new option `--coq-namespace` that allows specifying the prefix for the needed `Require Import` directives and it works without problem.

**Comment by jhjourdan** [comment]
Great. Could you do a PR for Menhir ?

**Comment by bschommer** [comment]
I will clean it up a little bit and probably submit it tomorrow.

**Comment by bschommer** [comment]
I opened an issue on the menhir gitlab and attached a simple patch.

**Comment by bschommer** [comment]
I created an PR with fixed `Require Imports ...` and a fixed `Makefile`

**Comment by xavierleroy** [comment]
Commit [master f3c60be] addresses the second part of this issue: fixing the 'Require Import' and 'Require Export' of Clightdefs.v and of clightgen-produced .v files so that they can be used in "strict name space mode", namely with
```
   -Q /path/to/compiled/compcert compcert
```
instead of CompCert's `-R` options.

Please review and let me know if this suits VST better.

If it can be useful for VST or other projects, we could add an `installcoqlib` target in the top-level Makefile that installs CompCert's .vo file hierarchy in a directory of the user's choice, ready to be used with `-Q /path/to/installed/compcert compcert`.

For the first part of this issue, namely using `-Q` instead of `-R` to build CompCert itself, I'm not completely convinced.  For one thing, Flocq will remain compiled with `-R` because rewriting imports inside Flocq's source files would make them diverge from the upstream distribution of Flocq.  The same goes for Menhir-related files at least until the Menhir folks want to put more name spacing and release and repackage a corresponding Menhir.  For another thing, no matter which syntax I try (`Require foo.bar`, `From foo Require Bar`), I feel that qualifying imports just adds noise to CompCert's sources but doesn't improve them.

**Comment by bschommer** [comment]
It seems that the issue is fixed at the moment. Please Reopen if the fix is not sufficient.

**Comment by andrew-appel** [comment]
This fix is not sufficient.  Consider this from the point of view of a user of VST.  He must access some things from CompCert (such as the syntactic operators of Clight statements and expressions, and several of the Flocq files) and some things from VST.

Because VST is built with -Q, and every import within VST is of the form,  "Require Import VST.foo.bar", then the hundreds of VST files do not pollute the user's namespace.

But the user must also Import Clight.  Because compcert is built with -R instead of -Q, that means (I believe) that the user must import compcert with -R instead of -Q, which means compcert pollutes the user's namespace with many dozens of top-level module names (file names), unqualified.

I understand @xavierleroy's point that one cannot easily fix CompCert without also fixing Flocq, but this just illustrates that Flocq is also broken in the same way: it pollutes the user's namespace.

However, I disagree with @xavierleroy's sentiment that "qualifying imports just adds noise to CompCert's sources but doesn't improve them."  The improvement is in their usability by others, when CompCert itself is used as a library.

**Comment by bschommer** [comment]
Then let's reopen it.

**Comment by bschommer** [comment]
One step would be to convince the authors of Flocq to compile their library with -Q. As @xavierleroy noted it is not desired to diverge from the upstream version.

**Comment by xavierleroy** [comment]
> Because compcert is built with -R instead of -Q, that means (I believe) that the user must import compcert with -R instead of -Q

You believe or you have evidence?  My experimental evidence is as follows:
* I installed the .vo files of CompCert in /path/to/compcert/lib/coq  using the new configure/Makefile support for installing the .vo files.
* I ran `coqide -Q /path/to/compcert/lib/coq`.
* `Require Import Coqlib` fails. (Like it should.)
* `Require Import lib.Coqlib` fails. (Like it should.)
* `Require Import compcert.lib.Coqlib` works.
* `From compcert Require Import  Coqlib` works.
* A .v file produced by clightgen loads fine.

I don't see any name space pollution here.  Can we please agree that the problem is solved and move on?


*

**Comment by xavierleroy** [comment]
(Typo in previous message, see Github for edited message.)

**Comment by andrew-appel** [comment]
Thank you, @xavierleroy.  Following your suggestion, this VST pull request
https://github.com/PrincetonUniversity/VST/pull/283
demonstrates that we can import CompCert into VST using -Q.
Therefore I can close this issue report.
